{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 1 - Introduction to Python and the Jupyter Notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this course we will make extensive use of Jupyter notebooks. Jupyter notebooks offer an interactive python session inside a browser, with great added functionality such as markdown support, widgets and much more. They are a great tool for data exploration, and freqently used in data science.\n",
    "\n",
    "To find out more:\n",
    "*http://jupyter.org/*\n",
    "\n",
    "This site has numerous examples and tutorials:\n",
    "*https://github.com/jupyter/jupyter/wiki/A-gallery-of-interesting-Jupyter-Notebooks*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Jupyter in the LEARN environment\n",
    "\n",
    "We are currently running a Python3 Jupyter Notebook within the LEARN system. This means there is no need for you to set Jupyter up on your own system you can do everything that you need for the upcoming lectures inside here. You may chose to install it locally for use during your coursework. If so you should first need to [install Python](https://www.python.org/downloads/) and then visit the [Jupyter Install page](http://jupyter.org/install) and follow the instructions there."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notebook Basics\n",
    "\n",
    "The notebook is organised in cells, each of which can hold markdown text (as this one) or python code. To execute the code in a cell, press the key combination ``Shift`` + ``Return`` or click the \"Run\" button in the control bar along the top of the Notebook.\n",
    "\n",
    "You can try this in the cell below, where we assign a string to the variable ``a`` and then print it:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = 'ATG'\n",
    "print(a)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once the cell above has been executed, the interpreter knows about ``a``, so we can start woking with it in the python way:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "print(a+a)\n",
    "print(a+a[::-1])\n",
    "print(len(a))\n",
    "print(len(a)**2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Further Python Basics\n",
    "Using the same approach as we used for uploading this Python Notebook you can download some more notebooks that introduce a range of basic python concepts from the [Python-Lectures](https://github.com/rajathkmp/Python-Lectures/blob/master/01.ipynb) GitHub site."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Biopython\n",
    "\n",
    "*http://biopython.org/*\n",
    "\n",
    "Biopython provides tools for analysis of genomics and proteomic data. We will use this throughout the course, so make sure this runs on your computer.\n",
    "\n",
    "First we need to install the Biopython package in your environment. **You only need to do this once**, it will remain in your account space througout the course."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting biopython\n",
      "  Downloading biopython-1.81.tar.gz (19.3 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m19.3/19.3 MB\u001b[0m \u001b[31m9.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0mm0:01\u001b[0mm\n",
      "\u001b[?25h  Installing build dependencies ... \u001b[?25ldone\n",
      "\u001b[?25h  Getting requirements to build wheel ... \u001b[?25ldone\n",
      "\u001b[?25h  Preparing metadata (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25hCollecting numpy (from biopython)\n",
      "  Obtaining dependency information for numpy from https://files.pythonhosted.org/packages/b4/ca/fc1c4f8a2a4693ff437d039acf2dc93a190b9494569fbed246f535c44fc8/numpy-1.26.0-cp39-cp39-macosx_11_0_arm64.whl.metadata\n",
      "  Downloading numpy-1.26.0-cp39-cp39-macosx_11_0_arm64.whl.metadata (53 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m53.5/53.5 kB\u001b[0m \u001b[31m7.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading numpy-1.26.0-cp39-cp39-macosx_11_0_arm64.whl (14.0 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m14.0/14.0 MB\u001b[0m \u001b[31m15.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hBuilding wheels for collected packages: biopython\n",
      "  Building wheel for biopython (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for biopython: filename=biopython-1.81-cp39-cp39-macosx_10_9_universal2.whl size=2823180 sha256=e8543eadf54b6448a94a0515ce3870dbb4ee0bac0ba2b27867986e2941700f09\n",
      "  Stored in directory: /Users/kuanlon/Library/Caches/pip/wheels/f3/1c/e0/95efa51f6855c29d87d42076680d0405078bd061f46d775488\n",
      "Successfully built biopython\n",
      "Installing collected packages: numpy, biopython\n",
      "Successfully installed biopython-1.81 numpy-1.26.0\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "### Install the biopython package in your Noteable environment\n",
    "%pip install biopython\n",
    "\n",
    "import Bio"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can find an excellent [biopython cookbook](http://biopython.org/DIST/docs/tutorial/Tutorial.html) written by the biopython community in the resource list for this course that you can practice with to familiarise yourself with some of its functionality."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create a DNA sequence using BioPython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGTACACTGGT\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "my_seq = Seq(\"AGTACACTGGT\")\n",
    "print(my_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Basic operations on DNA sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11 nucletotides long\n",
      "%GC content =  45.45454545454545 %\n",
      "%GC content = 45.45%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kuanlon/Documents/university/bioinformatics/bioinformatics1/.venv/lib/python3.9/site-packages/Bio/SeqUtils/__init__.py:144: BiopythonDeprecationWarning: GC is deprecated; please use gc_fraction instead.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "#sequence length\n",
    "print(len(my_seq),\"nucletotides long\")\n",
    "\n",
    "#sequence %GC content\n",
    "from Bio.SeqUtils import GC\n",
    "\n",
    "#simple print\n",
    "print(\"%GC content = \",GC(my_seq),\"%\")\n",
    "\n",
    "#printing to two decimal places\n",
    "print(\"%GC content = \"+'%4.2f' % GC(my_seq)+\"%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "original sequence AGTACACTGGT\n",
      "indexing from 1->5 GTAC\n",
      "indexing from 0->5 AGTAC\n"
     ]
    }
   ],
   "source": [
    "#original sequence\n",
    "print(\"original sequence\",my_seq)\n",
    "\n",
    "#sequence slicing NB this displays nucleotides 2-5\n",
    "print(\"indexing from 1->5\",my_seq[1:5])\n",
    "\n",
    "#the sequence is indexed from 0\n",
    "print(\"indexing from 0->5\",my_seq[0:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGTAGGGGGGGGGGCACTGGGGGGGGGGGGT\n"
     ]
    }
   ],
   "source": [
    "fragments = [Seq(\"AGTA\"), Seq(\"CACT\"), Seq(\"GGT\")]\n",
    "filler = Seq(\"G\"*10)\n",
    "print(filler.join(fragments))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find the starting index of a subsequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAC index: 4\n"
     ]
    }
   ],
   "source": [
    "seq_example = Seq(\"AGTACACTGGT\")\n",
    "print(\"CAC index:\", seq_example.find(\"CAC\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGTACACTGGT\n",
      "TCATGTGACCA\n",
      "ACCAGTGTACT\n"
     ]
    }
   ],
   "source": [
    "print(my_seq)\n",
    "\n",
    "#complement of the sequence\n",
    "print(my_seq.complement())\n",
    "\n",
    "#reverse complement of the sequence\n",
    "print(my_seq.reverse_complement())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Biopython contains useful meta-data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 1 Standard, SGC0\n",
      "\n",
      "  |  T      |  C      |  A      |  G      |\n",
      "--+---------+---------+---------+---------+--\n",
      "T | TTT F   | TCT S   | TAT Y   | TGT C   | T\n",
      "T | TTC F   | TCC S   | TAC Y   | TGC C   | C\n",
      "T | TTA L   | TCA S   | TAA Stop| TGA Stop| A\n",
      "T | TTG L(s)| TCG S   | TAG Stop| TGG W   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "C | CTT L   | CCT P   | CAT H   | CGT R   | T\n",
      "C | CTC L   | CCC P   | CAC H   | CGC R   | C\n",
      "C | CTA L   | CCA P   | CAA Q   | CGA R   | A\n",
      "C | CTG L(s)| CCG P   | CAG Q   | CGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "A | ATT I   | ACT T   | AAT N   | AGT S   | T\n",
      "A | ATC I   | ACC T   | AAC N   | AGC S   | C\n",
      "A | ATA I   | ACA T   | AAA K   | AGA R   | A\n",
      "A | ATG M(s)| ACG T   | AAG K   | AGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "G | GTT V   | GCT A   | GAT D   | GGT G   | T\n",
      "G | GTC V   | GCC A   | GAC D   | GGC G   | C\n",
      "G | GTA V   | GCA A   | GAA E   | GGA G   | A\n",
      "G | GTG V   | GCG A   | GAG E   | GGG G   | G\n",
      "--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "from Bio.Data import CodonTable\n",
    "standard_table = CodonTable.unambiguous_dna_by_id[1]\n",
    "\n",
    "print(standard_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TAA', 'TAG', 'TGA']\n"
     ]
    }
   ],
   "source": [
    "#and STOP codons\n",
    "print(standard_table.stop_codons)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Challenge 1 - Creating a Random Sequence\n",
    "An **optional** challenge to get a bit of basic practice manipulating sequences in BioPython. I will post example answers on the course Discussion Board before next week's class.\n",
    "\n",
    "Create a random DNA sequence of length 100 base pairs and print it out. HINT you will need to use the Python ``random`` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('CCGGTGCGGTCTGCTGCCCCGTCTGAGTCCGCGCTATGCTAGCAGCATACCGAG...ATG')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "random.seed(0)\n",
    "\n",
    "bases = ['A', 'G', 'T', 'C']\n",
    "ran_seq = Seq(\"\".join(random.choices(bases, k = 100)))\n",
    "\n",
    "ran_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Challenge 2 - Creating Mutated Sequences\n",
    "\n",
    "Another **optional** challenge, again I will post a possible answer on the course Discussion Board before next week's class. First I will help out by introducing some functions that you will probably want to use. There are many ways to do this so there is no one correct answer!\n",
    "\n",
    "Using the random sequence you created above, make 20 random mutations in it replacing the original base with a random lower case one. Then print the original sequence with the mutated sequence below it.\n",
    "\n",
    "HINT 1 - You will need to convert your random sequence into a ``MutableSeq`` object, see the BioPython cookbook for an example of it in use.\n",
    "\n",
    "HINT 2 - You will want to select a random position in the sequence, you'll probably want to use the ``random.randrange()`` function for that.\n",
    "\n",
    "Your final ouput should look like this:-\n",
    "\n",
    "```\n",
    "GTAAGCGCGTTGGGTTTGAAAGCCCACCGCAAAATGAAGCTCTAAGCAAACTGGGATAAATTGGCGACCCCGCACTGTTAGGACCGAAAGGTTTGTGACA\n",
    "cTgcGCcCGTTGGGTTTGcAcGCCCACtGtAAAATGAAGttCTAAGCAAACTGGGATAAATTGtCGtCtCCGCACTGTTgGGACCGgAAGGTTTGtGcCA\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCGGTGCGGTCTGCTGCCCCGTCTGAGTCCGCGCTATGCTAGCAGCATACCGAGTCATTTCTCTTGTGTGAATTGACCCCCCTGTGCCCTCTGTCCCATG\n",
      "CCagcGCGGaCcGCaaCCCCGTCaaAgTCCGCGCTATGCTAGCAGCAtACgGAGTCATTTCtCTTGTGaGAATgGACaCCCtTGTGCcCTCcGgCCCATG\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import MutableSeq\n",
    "\n",
    "mut_seq = MutableSeq(ran_seq)\n",
    "\n",
    "ran_positions = random.sample(range(100), 20)\n",
    "\n",
    "lc_bases = ['a','t', 'g', 'c']\n",
    "\n",
    "for i in ran_positions:\n",
    "    temp_base = random.choice(lc_bases)\n",
    "    mut_seq[i] = temp_base\n",
    "\n",
    "print(ran_seq)\n",
    "print(mut_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Matplotlib"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is essentailly a library for data ploting, and is useful to use in combination with ``numpy``. Try this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<matplotlib.lines.Line2D at 0x1165f7310>]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAiMAAAGdCAYAAADAAnMpAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAkjUlEQVR4nO3dbWxb5R338Z/tNjGI1BBKYreL1FCNByvQrC3JwoM0ppRGSIG+QMtNaWEVtFooiDWaRAulJjw0HWOot0RJRQGBVFirIdgIVC5bNl4wIkVqqESWUgQNalbspCXCiYrSgH3uF7nj1sRu4zz48rG/H+m8yMl17H84gH8+15PDsixLAAAAhjhNFwAAAPIbYQQAABhFGAEAAEYRRgAAgFGEEQAAYBRhBAAAGEUYAQAARhFGAACAUXNMFzAZsVhM33zzjYqKiuRwOEyXAwAAJsGyLA0PD2vBggVyOlM//7BFGPnmm29UVlZmugwAADAFfX19+tnPfpby97YII0VFRZLG/ph58+YZrgYAAEzG0NCQysrK4p/jqdgijIx3zcybN48wAgCAzVxoiAUDWAEAgFGEEQAAYBRhBAAAGEUYAQAARhFGAACAUYQRAABgFGEEAAAYRRgBAABG2WLRMwAATIvGLHX2DmpgeEQlRW5VlRfL5WS/tJlAGAEA4AKC3SE1t/UoFBmJn/N53ArU+1VX4TNYWW6gmwYAgPMIdofUuLcrIYhIUjgyosa9XQp2hwxVljsIIwAApBCNWWpu65GV5Hfj55rbehSNJWuBySKMAACQQmfv4IQnIueyJIUiI+rsHcxcUTmIMAIAQAoDw6mDyFTaITnCCAAAKZQUuWe0HZJjNs0sYQoYANhfVXmxfB63wpGRpONGHJK8nrH/x2PqCCOzgClgAJAbXE6HAvV+Ne7tkkNKCCTjXy8D9X6+bE4T3TQzjClgAJBb6ip8al2zVF5PYleM1+NW65qlfMmcATwZmUEXmgLm0NgUsBV+LykaAGykrsKnFX4v3e+zhDAyg9KZAlaz+PLMFQYAmDaX08H/u2cJ3TQziClgAACkjzAyg5gCBgBA+ggjM2h8CliqHkSHxmbVMAUMAICzCCMzaHwKmKQJgYQpYAAAJEcYmWFMAQMAID3MppkFTAEDAGDyCCOzhClgAABMDt00AADAKMIIAAAwijACAACMIowAAACjCCMAAMAowggAADCKMAIAAIwijAAAAKMIIwAAwCjCCAAAMIowAgAAjCKMAAAAowgjAADAKMIIAAAwijACAACMIowAAACjCCMAAMCoOaYLADB10Zilzt5BDQyPqKTIraryYrmcDtNlAUBaCCOATQW7Q2pu61EoMhI/5/O4Faj3q67CZ7AyAEgP3TSADQW7Q2rc25UQRCQpHBlR494uBbtDhioDgPQRRgCbicYsNbf1yEryu/FzzW09isaStQCA7EMYAWyms3dwwhORc1mSQpERdfYOZq4oAJgGwghgMwPDqYPIVNoBgGmEEcBmSorcM9oOAEwjjAA2U1VeLJ/HrVQTeB0am1VTVV6cybIAYMoII4DNuJwOBer9kjQhkIz/HKj3s94IANsgjAA2VFfhU+uapfJ6ErtivB63WtcsZZ0RALbComeATdVV+LTC72UFVgC2RxgBbMzldKhm8eWmywCAaaGbBgAAGEUYAQAARtFNg6zBDrQAkJ8II8gK7EALAPmLbhoYxw60AJDfCCMwih1oAQBTCiO7du3SokWL5Ha7VV1drc7OzvO237lzp66++mpddNFFKisr06ZNmzQywiZeYAdaAMAUwsj+/fvV1NSkQCCgrq4uLVmyRCtXrtTAwEDS9m+99ZY2b96sQCCgI0eO6NVXX9X+/fv12GOPTbt42B870AIA0g4jL7zwgtavX69169bJ7/dr9+7duvjii/Xaa68lbf/JJ5/opptu0urVq7Vo0SLddtttuvvuuy/4NAX5gR1oAQBphZHR0VEdOnRItbW1Z1/A6VRtba06OjqSXnPjjTfq0KFD8fBx7NgxHThwQLfffnvK9zlz5oyGhoYSDuQmdqAFAKQVRk6dOqVoNKrS0tKE86WlpQqHw0mvWb16tZ566indfPPNmjt3rhYvXqxf/epX5+2maWlpkcfjiR9lZWXplAkbYQdaAMCsz6b56KOPtH37dr300kvq6urSO++8ow8++EBPP/10ymu2bNmiSCQSP/r6+ma7TBjEDrQAkN/SWvRs/vz5crlc6u/vTzjf398vr9eb9JonnnhCa9eu1QMPPCBJuu6663T69Glt2LBBjz/+uJzOiXmosLBQhYWF6ZQGm2MHWgDIX2k9GSkoKNCyZcvU3t4ePxeLxdTe3q6ampqk13z//fcTAofL5ZIkWRZrR+Cs8R1o76xcqJrFlxNEACBPpL0cfFNTk+677z4tX75cVVVV2rlzp06fPq1169ZJku69914tXLhQLS0tkqT6+nq98MIL+sUvfqHq6mp9+eWXeuKJJ1RfXx8PJQAAIH+lHUYaGhp08uRJbdu2TeFwWJWVlQoGg/FBrcePH094ErJ161Y5HA5t3bpVJ06c0BVXXKH6+no9++yzM/dXAAAA23JYNugrGRoaksfjUSQS0bx580yXAwAAJmGyn9/sTQMAAIwijAAAAKMIIwAAwCjCCAAAMIowAgAAjCKMAAAAowgjAADAKMIIAAAwijACAACMIowAAACjCCMAAMAowggAADCKMAIAAIwijAAAAKMIIwAAwCjCCAAAMIowAgAAjCKMAAAAowgjAADAKMIIAAAwijACAACMIowAAACjCCMAAMAowggAADCKMAIAAIwijAAAAKMIIwAAwCjCCAAAMIowAgAAjCKMAAAAowgjAADAKMIIAAAwijACAACMIowAAACj5pguIJ9FY5Y6ewc1MDyikiK3qsqL5XI6TJcFAEBGEUYMCXaH1NzWo1BkJH7O53ErUO9XXYXPYGUAAGQW3TQGBLtDatzblRBEJCkcGVHj3i4Fu0OGKgMAIPMIIxkWjVlqbuuRleR34+ea23oUjSVrAQBA7iGMZFhn7+CEJyLnsiSFIiPq7B3MXFEAABhEGMmwgeHUQWQq7QAAsDvCSIaVFLlntB0AAHZHGMmwqvJi+TxupZrA69DYrJqq8uJMlgUAgDGEkQxzOR0K1PslaUIgGf85UO9nvREAQN4gjBhQV+FT65ql8noSu2K8Hrda1yxlnREAQF5h0TND6ip8WuH3sgIrACDvEUYMcjkdqll8uekygAti6wIAs4kwAuC82LoAwGxjzAiAlNi6AEAmEEYAJMXWBQAyhTACICm2LgCQKYQRAEmxdQGATCGMAEiKrQsAZAphBEBSbF0AIFMIIwCSYusCAJlCGAGQElsXAMgEFj3LMFayhN2wdQGA2UYYySBWsoRdsXUBgNlEN02GsJIlAADJEUYygJUsAQBIjTCSAaxkCQBAaoSRDGAlSwAAUiOMZAArWQIAkBphJANYyRIAgNSmFEZ27dqlRYsWye12q7q6Wp2dnedt/91332njxo3y+XwqLCzUVVddpQMHDkypYDtiJUsAAFJLO4zs379fTU1NCgQC6urq0pIlS7Ry5UoNDAwkbT86OqoVK1bo66+/1ttvv62jR49qz549Wrhw4bSLtxNWsgQAIDmHZVlpzSetrq7WDTfcoBdffFGSFIvFVFZWpocfflibN2+e0H737t3605/+pM8//1xz586dUpFDQ0PyeDyKRCKaN2/elF4jW7ACKwAgX0z28zutJyOjo6M6dOiQamtrz76A06na2lp1dHQkvea9995TTU2NNm7cqNLSUlVUVGj79u2KRqMp3+fMmTMaGhpKOHLF+EqWd1YuVM3iywkiAIC8l1YYOXXqlKLRqEpLSxPOl5aWKhwOJ73m2LFjevvttxWNRnXgwAE98cQT+vOf/6xnnnkm5fu0tLTI4/HEj7KysnTKBAAANjLrs2lisZhKSkr08ssva9myZWpoaNDjjz+u3bt3p7xmy5YtikQi8aOvr2+2ywQAAIaktVHe/Pnz5XK51N/fn3C+v79fXq836TU+n09z586Vy+WKn7v22msVDoc1OjqqgoKCCdcUFhaqsLAwndIAAIBNpfVkpKCgQMuWLVN7e3v8XCwWU3t7u2pqapJec9NNN+nLL79ULBaLn/viiy/k8/mSBhEAAJBf0u6maWpq0p49e/TGG2/oyJEjamxs1OnTp7Vu3TpJ0r333qstW7bE2zc2NmpwcFCPPPKIvvjiC33wwQfavn27Nm7cOHN/BQAAsK20umkkqaGhQSdPntS2bdsUDodVWVmpYDAYH9R6/PhxOZ1nM05ZWZkOHjyoTZs26frrr9fChQv1yCOP6NFHH525vwIAANhW2uuMmJBL64wAAJAvZmWdEQAAgJlGGAEAAEYRRgAAgFGEEQAAYBRhBAAAGEUYAQAARhFGAACAUYQRAABgFGEEAAAYlfZy8AAAIPtFY5Y6ewc1MDyikiK3qsqL5XI6TJeVFGEEAIAcE+wOqbmtR6HISPycz+NWoN6vugqfwcqSo5sGAFKIxix1fPWt/n74hDq++lbRWNZv5QUo2B1S496uhCAiSeHIiBr3dinYHTJUWWo8GQGAJOz2zRKQxgJ0c1uPksVmS5JDUnNbj1b4vVnVZcOTEQD4CTt+swQkqbN3cMK/t+eyJIUiI+rsHcxcUZNAGAGAc1zom6U09s2SLhtko4Hh1EFkKu0yhTACAOew6zdLQJJKitwz2i5TCCMAcA67frMEJKmqvFg+j1upRoM4NDb2qaq8OJNlXRBhBADOYddvloAkuZwOBer9kjQhkIz/HKj3Z9XgVYkwAgAJ7PrNEhhXV+FT65ql8noSA7PX41brmqVZORuMqb0AcI7xb5aNe7vkkBIGsmbzN0vgXHUVPq3we22zAqvDsqysHxI+NDQkj8ejSCSiefPmmS4HQB5gnRFg+ib7+c2TEdiKnfZagL3Z7ZslYGeEEdgG31SRaS6nQzWLLzddBpDzGMAKW2BFTADIXYQRZD1WxASQD/J5Y0a6aZD10lkRk0fqAOwo37uheTKCrMeKmAByGd3QhBHYACtiAshVdEOPIYwg67EiJoBcxcaMYwgjyHp23WsBAC6EbugxhBHYgh33WgCAC6EbegyzaWAbrIgJINeMd0OHIyNJx404NPalK9e7oQkjsBVWxASQS9iYcQzdNAAAGEQ3NE9GAAAwLt+7oQkjAABkgXzuhqabBgAAGEUYAQAARhFGAACAUYQRAABgFGEEAAAYRRgBAABGEUYAAIBRhBEAAGAUYQQAABhFGAEAAEYRRgAAgFHsTQMAMyAas/J2kzNguggjADBNwe6Qmtt6FIqMxM/5PG4F6v15sf07MF100wDANAS7Q2rc25UQRCQpHBlR494uBbtDhioD7IMwAgBTFI1Zam7rkZXkd+Pnmtt6FI0lawFgHGEEAKaos3dwwhORc1mSQpERdfYOZq4owIYIIwAwRQPDqYPIVNoB+YowAgBTVFLkntF2QL4ijADAFFWVF8vncSvVBF6HxmbVVJUXZ7IswHYIIwAwRS6nQ4F6vyRNCCTjPwfq/aw3AlwAYQQApqGuwqfWNUvl9SR2xXg9brWuWco6I8AksOgZAExTXYVPK/xeVmAFpogwAgAzwOV0qGbx5abLAGyJbhoAAGAUYQQAABhFGAEAAEYRRgAAgFFTCiO7du3SokWL5Ha7VV1drc7Ozkldt2/fPjkcDq1atWoqbwsAAGZQNGap46tv9ffDJ9Tx1bfGNnVMezbN/v371dTUpN27d6u6ulo7d+7UypUrdfToUZWUlKS87uuvv9Yf/vAH3XLLLdMqGAAATF+wO6Tmtp6EzR59HrcC9f6Mr4+T9pORF154QevXr9e6devk9/u1e/duXXzxxXrttddSXhONRnXPPfeoublZV1555bQKBgAA0xPsDqlxb9eEXafDkRE17u1SsDuU0XrSCiOjo6M6dOiQamtrz76A06na2lp1dHSkvO6pp55SSUmJ7r///km9z5kzZzQ0NJRwAACA6YvGLDW39ShZh8z4uea2nox22aQVRk6dOqVoNKrS0tKE86WlpQqHw0mv+fjjj/Xqq69qz549k36flpYWeTye+FFWVpZOmQAAIIXO3sEJT0TOZUkKRUbU2TuYsZpmdTbN8PCw1q5dqz179mj+/PmTvm7Lli2KRCLxo6+vbxarBADYTbYMvLSjgeHUQWQq7WZCWgNY58+fL5fLpf7+/oTz/f398nq9E9p/9dVX+vrrr1VfXx8/F4vFxt54zhwdPXpUixcvnnBdYWGhCgsL0ykNAJAnsmngpR2VFLkv3CiNdjMhrScjBQUFWrZsmdrb2+PnYrGY2tvbVVNTM6H9Nddco88++0yHDx+OH3fccYduvfVWHT58mO4XAEBasm3gpR1VlRfL53Er1TaODo2Fu6ry4ozVlPbU3qamJt13331avny5qqqqtHPnTp0+fVrr1q2TJN17771auHChWlpa5Ha7VVFRkXD9pZdeKkkTzgMAcD4XGnjp0NjAyxV+b07umByNWTOyM7TL6VCg3q/GvV1ySAn/PMdfLVDvz+g/w7TDSENDg06ePKlt27YpHA6rsrJSwWAwPqj1+PHjcjpZ2BUAMLPSGXiZazsoz3TXVF2FT61rlk54Ta+h7i6HZVlZP+pnaGhIHo9HkUhE8+bNM10OAMCAvx8+oUf2Hb5gu//7fyp1Z+XC2S8oQ8a7pn76YT3+3KJ1zdIph4eZetqSymQ/v9N+MgIAgAnZOPByts1215TL6ciKp0j0pwAAbCEbB17OtmxcE2Q2EEYAALYwPvBS0oRAYmrg5WzLxjVBZgNhBABgG+MDL72exK4Yr8c9rbET2SpfuqYYMwIAsJW6Cp9W+L2zOvAyW4x3TYUjI0nHjTg0FsTs3jVFGAEA2E62DLycbdm4JshsoJsGAIAslg9dUzwZAQAgy+V61xRhBAAAG8jlrim6aQAAgFGEEQAAYBRhBAAAGEUYAQAARhFGAACAUYQRAABgFFN7ASCHRWNWzq5NgdxBGAGAHBXsDqm5rSdhC3qfx61AvT8nVu1E7qCbBgByULA7pMa9XQlBRJLCkRE17u1SsDtkqDJgIsIIAOSYaMxSc1tP0l1ex881t/UoGkvWAsg8wggA5JjO3sEJT0TOZUkKRUbU2TuYuaKA8yCMAECOGRhOHUSm0g6YbYQRAMgxJUXuCzdKox0w2wgjAJBjqsqL5fO4lWoCr0Njs2qqyoszWRaQEmEEAHKMy+lQoN4vSRMCyfjPgXo/640gaxBGACAH1VX41LpmqbyexK4Yr8et1jVLWWcEWYVFzwAgR9VV+LTC72UFVmQ9wggA5DCX06GaxZebLgM4L7ppAACAUYQRAABgFGEEAAAYRRgBAABGEUYAAIBRhBEAAGAUYQQAABjFOiNADovGLBa8ApD1CCNAjgp2h9Tc1qNQ5Ow28T6PW4F6P0uBA8gqdNMAOSjYHVLj3q6EICJJ4ciIGvd2KdgdMlQZAExEGAFyTDRmqbmtR1aS342fa27rUTSWrAUAZB5hBMgxnb2DE56InMuSFIqMqLN3MHNFAcB5EEaAHDMwnDqITKUdAMw2wgiQY0qK3DPaDgBmG2EEyDFV5cXyedxKNYHXobFZNVXlxZksCwBSIowAOcbldChQ75ekCYFk/OdAvZ/1RgBkDcIIkIPqKnxqXbNUXk9iV4zX41brmqWsMwIgq7DoGZCj6ip8WuH3sgIrgKxHGAFymMvpUM3iy02XAQDnRTcNAAAwijACAACMIowAAACjCCMAAMAowggAADCKMAIAAIwijAAAAKMIIwAAwCjCCAAAMIowAgAAjCKMAAAAowgjAADAKMIIAAAwijACAACMmmO6AADTE41Z6uwd1MDwiEqK3KoqL5bL6TBdFgBMGmEEsLFgd0jNbT0KRUbi53wetwL1ftVV+AxWBgCTRzcNYFPB7pAa93YlBBFJCkdG1Li3S8HukKHKACA9Uwoju3bt0qJFi+R2u1VdXa3Ozs6Ubffs2aNbbrlFl112mS677DLV1taetz2AC4vGLDW39chK8rvxc81tPYrGkrUAgOySdhjZv3+/mpqaFAgE1NXVpSVLlmjlypUaGBhI2v6jjz7S3XffrX//+9/q6OhQWVmZbrvtNp04cWLaxQP5qrN3cMITkXNZkkKREXX2DmauKACYorTDyAsvvKD169dr3bp18vv92r17ty6++GK99tprSdu/+eabevDBB1VZWalrrrlGr7zyimKxmNrb26ddPJCvBoZTB5GptAMAk9IKI6Ojozp06JBqa2vPvoDTqdraWnV0dEzqNb7//nv98MMPKi4uTtnmzJkzGhoaSjgAnFVS5J7RdgBgUlph5NSpU4pGoyotLU04X1paqnA4PKnXePTRR7VgwYKEQPNTLS0t8ng88aOsrCydMoGcV1VeLJ/HrVQTeB0am1VTVZ469ANAtsjobJodO3Zo3759evfdd+V2p/7GtmXLFkUikfjR19eXwSqB7OdyOhSo90vShEAy/nOg3s96IwBsIa0wMn/+fLlcLvX39yec7+/vl9frPe+1zz//vHbs2KEPP/xQ119//XnbFhYWat68eQkHgER1FT61rlkqrycx2Hs9brWuWco6IwBsI61FzwoKCrRs2TK1t7dr1apVkhQfjPrQQw+lvO65557Ts88+q4MHD2r58uXTKhjAWXUVPq3we1mBFYCtpb0Ca1NTk+677z4tX75cVVVV2rlzp06fPq1169ZJku69914tXLhQLS0tkqQ//vGP2rZtm9566y0tWrQoPrbkkksu0SWXXDKDfwqQn1xOh2oWX266DACYsrTDSENDg06ePKlt27YpHA6rsrJSwWAwPqj1+PHjcjrP9v60trZqdHRUd911V8LrBAIBPfnkk9OrHgAA2J7DsqysX6JxaGhIHo9HkUiE8SMAANjEZD+/2ZsGAAAYRRgBAABGEUYAAIBRhBEAAGAUYQQAABhFGAEAAEYRRgAAgFGEEQAAYBRhBAAAGEUYAQAARhFGAACAUYQRAABgFGEEAAAYRRgBAABGEUYAAIBRhBEAAGAUYQQAABhFGAEAAEYRRgAAgFGEEQAAYBRhBAAAGEUYAQAARhFGAACAUYQRAABgFGEEAAAYRRgBAABGEUYAAIBRhBEAAGAUYQQAABhFGAEAAEYRRgAAgFGEEQAAYBRhBAAAGEUYAQAARhFGAACAUYQRAABgFGEEAAAYRRgBAABGEUYAAIBRhBEAAGAUYQQAABhFGAEAAEYRRgAAgFGEEQAAYBRhBAAAGDXHdAGmRGOWOnsHNTA8opIit6rKi+VyOkyXBQBA3snLMBLsDqm5rUehyEj8nM/jVqDer7oKn8HKAADIP3nXTRPsDqlxb1dCEJGkcGREjXu7FOwOGaoMAID8lFdhJBqz1NzWIyvJ78bPNbf1KBpL1gIAAMyGvAojnb2DE56InMuSFIqMqLN3MHNFzbJozFLHV9/q74dPqOOrbwlaAICsk1djRgaGUweRqbTLdoyNAQDYQV49GSkpcs9ou2zG2BgAgF3kVRipKi+Wz+NWqgm8Do09OagqL85kWTOOsTEAADvJqzDicjoUqPdL0oRAMv5zoN5v+/VG8nFsDADAvvIqjEhSXYVPrWuWyutJ7IrxetxqXbM0J8ZS5NvYGACAveXVANZxdRU+rfB7c3YF1nwaGwMAsL+8DCPSWJdNzeLLTZcxK8bHxoQjI0nHjTg09iTI7mNjAAC5Ie+6afJBvoyNAQDkBsJIjsqHsTEAgNyQt900+SDXx8YAAHIDYSTH5fLYGABAbqCbBgAAGEUYAQAARk0pjOzatUuLFi2S2+1WdXW1Ojs7z9v+r3/9q6655hq53W5dd911OnDgwJSKBQAAuSftMLJ//341NTUpEAioq6tLS5Ys0cqVKzUwMJC0/SeffKK7775b999/vz799FOtWrVKq1atUnd397SLBwAA9uewLCut3dKqq6t1ww036MUXX5QkxWIxlZWV6eGHH9bmzZsntG9oaNDp06f1/vvvx8/98pe/VGVlpXbv3j2p9xwaGpLH41EkEtG8efPSKRcAABgy2c/vtJ6MjI6O6tChQ6qtrT37Ak6namtr1dHRkfSajo6OhPaStHLlypTtJenMmTMaGhpKOAAAQG5KK4ycOnVK0WhUpaWlCedLS0sVDoeTXhMOh9NqL0ktLS3yeDzxo6ysLJ0yAQCAjWTlbJotW7YoEonEj76+PtMlAQCAWZLWomfz58+Xy+VSf39/wvn+/n55vd6k13i93rTaS1JhYaEKCwvTKQ0AANhUWmGkoKBAy5YtU3t7u1atWiVpbABre3u7HnrooaTX1NTUqL29Xb///e/j5/7xj3+opqZm0u87PsaWsSMAANjH+Of2BefKWGnat2+fVVhYaL3++utWT0+PtWHDBuvSSy+1wuGwZVmWtXbtWmvz5s3x9v/5z3+sOXPmWM8//7x15MgRKxAIWHPnzrU+++yzSb9nX1+fJYmDg4ODg4PDhkdfX995P+fT3pumoaFBJ0+e1LZt2xQOh1VZWalgMBgfpHr8+HE5nWeHotx444166623tHXrVj322GP6+c9/rr/97W+qqKiY9HsuWLBAfX19KioqksPBJm+pDA0NqaysTH19fUyBtgnumT1x3+yJ+5Z5lmVpeHhYCxYsOG+7tNcZQfZiPRb74Z7ZE/fNnrhv2SsrZ9MAAID8QRgBAABGEUZySGFhoQKBANOibYR7Zk/cN3vivmUvxowAAACjeDICAACMIowAAACjCCMAAMAowggAADCKMGIzu3bt0qJFi+R2u1VdXa3Ozs6Ubffs2aNbbrlFl112mS677DLV1taetz1mRzr37Fz79u2Tw+GI7wOFzEr3vn333XfauHGjfD6fCgsLddVVV+nAgQMZqhbj0r1vO3fu1NVXX62LLrpIZWVl2rRpk0ZGRjJULeLS3ZsG5uzbt88qKCiwXnvtNeu///2vtX79euvSSy+1+vv7k7ZfvXq1tWvXLuvTTz+1jhw5Yv32t7+1PB6P9b///S/DleevdO/ZuN7eXmvhwoXWLbfcYt15552ZKRZx6d63M2fOWMuXL7duv/126+OPP7Z6e3utjz76yDp8+HCGK89v6d63N9980yosLLTefPNNq7e31zp48KDl8/msTZs2ZbhyEEZspKqqytq4cWP852g0ai1YsMBqaWmZ1PU//vijVVRUZL3xxhuzVSJ+Yir37Mcff7RuvPFG65VXXrHuu+8+wogB6d631tZW68orr7RGR0czVSKSSPe+bdy40fr1r3+dcK6pqcm66aabZrVOTEQ3jU2Mjo7q0KFDqq2tjZ9zOp2qra1VR0fHpF7j+++/1w8//KDi4uLZKhPnmOo9e+qpp1RSUqL7778/E2XiJ6Zy39577z3V1NRo48aNKi0tVUVFhbZv365oNJqpsvPeVO7bjTfeqEOHDsW7co4dO6YDBw7o9ttvz0jNOCvtXXthxqlTpxSNRuO7I48rLS3V559/PqnXePTRR7VgwYKE/1gxe6Zyzz7++GO9+uqrOnz4cAYqRDJTuW/Hjh3Tv/71L91zzz06cOCAvvzySz344IP64YcfFAgEMlF23pvKfVu9erVOnTqlm2++WZZl6ccff9Tvfvc7PfbYY5koGefgyUie2LFjh/bt26d3331XbrfbdDlIYnh4WGvXrtWePXs0f/580+UgDbFYTCUlJXr55Ze1bNkyNTQ06PHHH9fu3btNl4bz+Oijj7R9+3a99NJL6urq0jvvvKMPPvhATz/9tOnS8g5PRmxi/vz5crlc6u/vTzjf398vr9d73muff/557dixQ//85z91/fXXz2aZOEe69+yrr77S119/rfr6+vi5WCwmSZozZ46OHj2qxYsXz27RmNJ/az6fT3PnzpXL5Yqfu/baaxUOhzU6OqqCgoJZrRlTu29PPPGE1q5dqwceeECSdN111+n06dPasGGDHn/8cTmdfF/PFP5J20RBQYGWLVum9vb2+LlYLKb29nbV1NSkvO65557T008/rWAwqOXLl2eiVPx/6d6za665Rp999pkOHz4cP+644w7deuutOnz4sMrKyjJZft6ayn9rN910k7788st4eJSkL774Qj6fjyCSIVO5b99///2EwDEeKC22bcss0yNoMXn79u2zCgsLrddff93q6emxNmzYYF166aVWOBy2LMuy1q5da23evDnefseOHVZBQYH19ttvW6FQKH4MDw+b+hPyTrr37KeYTWNGuvft+PHjVlFRkfXQQw9ZR48etd5//32rpKTEeuaZZ0z9CXkp3fsWCASsoqIi6y9/+Yt17Ngx68MPP7QWL15s/eY3vzH1J+QtumlspKGhQSdPntS2bdsUDodVWVmpYDAYH7B1/PjxhJTf2tqq0dFR3XXXXQmvEwgE9OSTT2ay9LyV7j1Ddkj3vpWVlengwYPatGmTrr/+ei1cuFCPPPKIHn30UVN/Ql5K975t3bpVDodDW7du1YkTJ3TFFVeovr5ezz77rKk/IW85LItnUQAAwBy+kgEAAKMIIwAAwCjCCAAAMIowAgAAjCKMAAAAowgjAADAKMIIAAAwijACAACMIowAAACjCCMAAMAowggAADCKMAIAAIz6f29OujH5pxCDAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.plot(np.random.rand(20),np.random.rand(20),'o')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Python help system\n",
    "\n",
    "Python libraries and functions are well documented. To see help for a function easily from a cell you can type ``?function`` and execute the cell. This will show extended documentation at the bottom of the screen\n",
    "\n",
    "Execute this cell to try:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0;31mSignature:\u001b[0m\n",
      "\u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m'float | ArrayLike | str'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mscalex\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m'bool'\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mscaley\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m'bool'\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;34m'list[Line2D]'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mDocstring:\u001b[0m\n",
      "Plot y versus x as lines and/or markers.\n",
      "\n",
      "Call signatures::\n",
      "\n",
      "    plot([x], y, [fmt], *, data=None, **kwargs)\n",
      "    plot([x], y, [fmt], [x2], y2, [fmt2], ..., **kwargs)\n",
      "\n",
      "The coordinates of the points or line nodes are given by *x*, *y*.\n",
      "\n",
      "The optional parameter *fmt* is a convenient way for defining basic\n",
      "formatting like color, marker and linestyle. It's a shortcut string\n",
      "notation described in the *Notes* section below.\n",
      "\n",
      ">>> plot(x, y)        # plot x and y using default line style and color\n",
      ">>> plot(x, y, 'bo')  # plot x and y using blue circle markers\n",
      ">>> plot(y)           # plot y using x as index array 0..N-1\n",
      ">>> plot(y, 'r+')     # ditto, but with red plusses\n",
      "\n",
      "You can use `.Line2D` properties as keyword arguments for more\n",
      "control on the appearance. Line properties and *fmt* can be mixed.\n",
      "The following two calls yield identical results:\n",
      "\n",
      ">>> plot(x, y, 'go--', linewidth=2, markersize=12)\n",
      ">>> plot(x, y, color='green', marker='o', linestyle='dashed',\n",
      "...      linewidth=2, markersize=12)\n",
      "\n",
      "When conflicting with *fmt*, keyword arguments take precedence.\n",
      "\n",
      "\n",
      "**Plotting labelled data**\n",
      "\n",
      "There's a convenient way for plotting objects with labelled data (i.e.\n",
      "data that can be accessed by index ``obj['y']``). Instead of giving\n",
      "the data in *x* and *y*, you can provide the object in the *data*\n",
      "parameter and just give the labels for *x* and *y*::\n",
      "\n",
      ">>> plot('xlabel', 'ylabel', data=obj)\n",
      "\n",
      "All indexable objects are supported. This could e.g. be a `dict`, a\n",
      "`pandas.DataFrame` or a structured numpy array.\n",
      "\n",
      "\n",
      "**Plotting multiple sets of data**\n",
      "\n",
      "There are various ways to plot multiple sets of data.\n",
      "\n",
      "- The most straight forward way is just to call `plot` multiple times.\n",
      "  Example:\n",
      "\n",
      "  >>> plot(x1, y1, 'bo')\n",
      "  >>> plot(x2, y2, 'go')\n",
      "\n",
      "- If *x* and/or *y* are 2D arrays a separate data set will be drawn\n",
      "  for every column. If both *x* and *y* are 2D, they must have the\n",
      "  same shape. If only one of them is 2D with shape (N, m) the other\n",
      "  must have length N and will be used for every data set m.\n",
      "\n",
      "  Example:\n",
      "\n",
      "  >>> x = [1, 2, 3]\n",
      "  >>> y = np.array([[1, 2], [3, 4], [5, 6]])\n",
      "  >>> plot(x, y)\n",
      "\n",
      "  is equivalent to:\n",
      "\n",
      "  >>> for col in range(y.shape[1]):\n",
      "  ...     plot(x, y[:, col])\n",
      "\n",
      "- The third way is to specify multiple sets of *[x]*, *y*, *[fmt]*\n",
      "  groups::\n",
      "\n",
      "  >>> plot(x1, y1, 'g^', x2, y2, 'g-')\n",
      "\n",
      "  In this case, any additional keyword argument applies to all\n",
      "  datasets. Also, this syntax cannot be combined with the *data*\n",
      "  parameter.\n",
      "\n",
      "By default, each line is assigned a different style specified by a\n",
      "'style cycle'. The *fmt* and line property parameters are only\n",
      "necessary if you want explicit deviations from these defaults.\n",
      "Alternatively, you can also change the style cycle using\n",
      ":rc:`axes.prop_cycle`.\n",
      "\n",
      "\n",
      "Parameters\n",
      "----------\n",
      "x, y : array-like or scalar\n",
      "    The horizontal / vertical coordinates of the data points.\n",
      "    *x* values are optional and default to ``range(len(y))``.\n",
      "\n",
      "    Commonly, these parameters are 1D arrays.\n",
      "\n",
      "    They can also be scalars, or two-dimensional (in that case, the\n",
      "    columns represent separate data sets).\n",
      "\n",
      "    These arguments cannot be passed as keywords.\n",
      "\n",
      "fmt : str, optional\n",
      "    A format string, e.g. 'ro' for red circles. See the *Notes*\n",
      "    section for a full description of the format strings.\n",
      "\n",
      "    Format strings are just an abbreviation for quickly setting\n",
      "    basic line properties. All of these and more can also be\n",
      "    controlled by keyword arguments.\n",
      "\n",
      "    This argument cannot be passed as keyword.\n",
      "\n",
      "data : indexable object, optional\n",
      "    An object with labelled data. If given, provide the label names to\n",
      "    plot in *x* and *y*.\n",
      "\n",
      "    .. note::\n",
      "        Technically there's a slight ambiguity in calls where the\n",
      "        second label is a valid *fmt*. ``plot('n', 'o', data=obj)``\n",
      "        could be ``plt(x, y)`` or ``plt(y, fmt)``. In such cases,\n",
      "        the former interpretation is chosen, but a warning is issued.\n",
      "        You may suppress the warning by adding an empty format string\n",
      "        ``plot('n', 'o', '', data=obj)``.\n",
      "\n",
      "Returns\n",
      "-------\n",
      "list of `.Line2D`\n",
      "    A list of lines representing the plotted data.\n",
      "\n",
      "Other Parameters\n",
      "----------------\n",
      "scalex, scaley : bool, default: True\n",
      "    These parameters determine if the view limits are adapted to the\n",
      "    data limits. The values are passed on to\n",
      "    `~.axes.Axes.autoscale_view`.\n",
      "\n",
      "**kwargs : `~matplotlib.lines.Line2D` properties, optional\n",
      "    *kwargs* are used to specify properties like a line label (for\n",
      "    auto legends), linewidth, antialiasing, marker face color.\n",
      "    Example::\n",
      "\n",
      "    >>> plot([1, 2, 3], [1, 2, 3], 'go-', label='line 1', linewidth=2)\n",
      "    >>> plot([1, 2, 3], [1, 4, 9], 'rs', label='line 2')\n",
      "\n",
      "    If you specify multiple lines with one plot call, the kwargs apply\n",
      "    to all those lines. In case the label object is iterable, each\n",
      "    element is used as labels for each set of data.\n",
      "\n",
      "    Here is a list of available `.Line2D` properties:\n",
      "\n",
      "    Properties:\n",
      "    agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "    alpha: scalar or None\n",
      "    animated: bool\n",
      "    antialiased or aa: bool\n",
      "    clip_box: `~matplotlib.transforms.BboxBase` or None\n",
      "    clip_on: bool\n",
      "    clip_path: Patch or (Path, Transform) or None\n",
      "    color or c: color\n",
      "    dash_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "    dash_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "    dashes: sequence of floats (on/off ink in points) or (None, None)\n",
      "    data: (2, N) array or two 1D arrays\n",
      "    drawstyle or ds: {'default', 'steps', 'steps-pre', 'steps-mid', 'steps-post'}, default: 'default'\n",
      "    figure: `~matplotlib.figure.Figure`\n",
      "    fillstyle: {'full', 'left', 'right', 'bottom', 'top', 'none'}\n",
      "    gapcolor: color or None\n",
      "    gid: str\n",
      "    in_layout: bool\n",
      "    label: object\n",
      "    linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "    linewidth or lw: float\n",
      "    marker: marker style string, `~.path.Path` or `~.markers.MarkerStyle`\n",
      "    markeredgecolor or mec: color\n",
      "    markeredgewidth or mew: float\n",
      "    markerfacecolor or mfc: color\n",
      "    markerfacecoloralt or mfcalt: color\n",
      "    markersize or ms: float\n",
      "    markevery: None or int or (int, int) or slice or list[int] or float or (float, float) or list[bool]\n",
      "    mouseover: bool\n",
      "    path_effects: list of `.AbstractPathEffect`\n",
      "    picker: float or callable[[Artist, Event], tuple[bool, dict]]\n",
      "    pickradius: float\n",
      "    rasterized: bool\n",
      "    sketch_params: (scale: float, length: float, randomness: float)\n",
      "    snap: bool or None\n",
      "    solid_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "    solid_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "    transform: unknown\n",
      "    url: str\n",
      "    visible: bool\n",
      "    xdata: 1D array\n",
      "    ydata: 1D array\n",
      "    zorder: float\n",
      "\n",
      "See Also\n",
      "--------\n",
      "scatter : XY scatter plot with markers of varying size and/or color (\n",
      "    sometimes also called bubble chart).\n",
      "\n",
      "Notes\n",
      "-----\n",
      "**Format Strings**\n",
      "\n",
      "A format string consists of a part for color, marker and line::\n",
      "\n",
      "    fmt = '[marker][line][color]'\n",
      "\n",
      "Each of them is optional. If not provided, the value from the style\n",
      "cycle is used. Exception: If ``line`` is given, but no ``marker``,\n",
      "the data will be a line without markers.\n",
      "\n",
      "Other combinations such as ``[color][marker][line]`` are also\n",
      "supported, but note that their parsing may be ambiguous.\n",
      "\n",
      "**Markers**\n",
      "\n",
      "=============   ===============================\n",
      "character       description\n",
      "=============   ===============================\n",
      "``'.'``         point marker\n",
      "``','``         pixel marker\n",
      "``'o'``         circle marker\n",
      "``'v'``         triangle_down marker\n",
      "``'^'``         triangle_up marker\n",
      "``'<'``         triangle_left marker\n",
      "``'>'``         triangle_right marker\n",
      "``'1'``         tri_down marker\n",
      "``'2'``         tri_up marker\n",
      "``'3'``         tri_left marker\n",
      "``'4'``         tri_right marker\n",
      "``'8'``         octagon marker\n",
      "``'s'``         square marker\n",
      "``'p'``         pentagon marker\n",
      "``'P'``         plus (filled) marker\n",
      "``'*'``         star marker\n",
      "``'h'``         hexagon1 marker\n",
      "``'H'``         hexagon2 marker\n",
      "``'+'``         plus marker\n",
      "``'x'``         x marker\n",
      "``'X'``         x (filled) marker\n",
      "``'D'``         diamond marker\n",
      "``'d'``         thin_diamond marker\n",
      "``'|'``         vline marker\n",
      "``'_'``         hline marker\n",
      "=============   ===============================\n",
      "\n",
      "**Line Styles**\n",
      "\n",
      "=============    ===============================\n",
      "character        description\n",
      "=============    ===============================\n",
      "``'-'``          solid line style\n",
      "``'--'``         dashed line style\n",
      "``'-.'``         dash-dot line style\n",
      "``':'``          dotted line style\n",
      "=============    ===============================\n",
      "\n",
      "Example format strings::\n",
      "\n",
      "    'b'    # blue markers with default shape\n",
      "    'or'   # red circles\n",
      "    '-g'   # green solid line\n",
      "    '--'   # dashed line with default color\n",
      "    '^k:'  # black triangle_up markers connected by a dotted line\n",
      "\n",
      "**Colors**\n",
      "\n",
      "The supported color abbreviations are the single letter codes\n",
      "\n",
      "=============    ===============================\n",
      "character        color\n",
      "=============    ===============================\n",
      "``'b'``          blue\n",
      "``'g'``          green\n",
      "``'r'``          red\n",
      "``'c'``          cyan\n",
      "``'m'``          magenta\n",
      "``'y'``          yellow\n",
      "``'k'``          black\n",
      "``'w'``          white\n",
      "=============    ===============================\n",
      "\n",
      "and the ``'CN'`` colors that index into the default property cycle.\n",
      "\n",
      "If the color is the only part of the format string, you can\n",
      "additionally use any  `matplotlib.colors` spec, e.g. full names\n",
      "(``'green'``) or hex strings (``'#008000'``).\n",
      "\u001b[0;31mFile:\u001b[0m      ~/Documents/university/bioinformatics/bioinformatics1/.venv/lib/python3.9/site-packages/matplotlib/pyplot.py\n",
      "\u001b[0;31mType:\u001b[0m      function"
     ]
    }
   ],
   "source": [
    "?plt.plot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Shutting Down the Notebook\n",
    "\n",
    "To shutdown the Notebook choose \"File->Close and Halt\" from the menu at the top of the Notebook.\n",
    "\n",
    "### Questions?\n",
    "\n",
    "Any questions feel free to put any quesitons on the Discussion Forum in the \"Tutorial Information & Discussion\" Channel."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.3 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "c79478e135452d4f8dcea3898ce85a4457be8d06848dc07bbec8d2854f4ceed7"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
